File: pepper_tools/setposture/setposture.py


import qi
import argparse
import sys
import os

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--pip", type=str, default=os.environ['PEPPER_IP'],
                        help="Robot IP address.  On robot or Local Naoqi: use '127.0.0.1'.")
    parser.add_argument("--pport", type=int, default=9559,
                        help="Naoqi port number")
    parser.add_argument("--posture", type=str, default="Stand",
                        help="Desired robot posture. Choose among: Stand, StandZero, Crouch")
    args = parser.parse_args()
    pip = args.pip
    pport = args.pport
    posture = args.posture

    #Start working session
    session = qi.Session()
    try:
        session.connect("tcp://" + pip + ":" + str(pport))
    except RuntimeError:
        print ("Can't connect to Naoqi at ip \"" + pip + "\" on port " + str(pport) +".\n"
               "Please check your script arguments. Run with -h option for help.")
        sys.exit(1)


    tts_service = session.service("ALTextToSpeech")
    rp_service = session.service("ALRobotPosture")

    current_posture = rp_service.getPosture()
    print "Robot posture: ", current_posture

    phraseToSay = "Hello! My current posture is " + current_posture
    tts_service.say(phraseToSay)

    if (current_posture != posture):
        phraseToSay = "Changing my posture to " + posture
        tts_service.say(phraseToSay)

        rp_service.goToPosture(posture,1.0)
        current_posture = rp_service.getPosture()
        print "Robot posture: ", current_posture
    else:
        phraseToSay = "Nothing to change here."
        tts_service.say(phraseToSay)







if __name__ == "__main__":
    main()

File: pepper_tools/setposture/setposture.cpp
/**
 * Copyright (c) 2011 Aldebaran Robotics. All Rights Reserved
 * \file sayhelloworld.cpp
 * \brief Make NAO say a short phrase.
 *
 * A simple example showing how to make NAO say a short phrase using the
 * specialized proxy ALTextToSpeechProxy.
 */


#include <iostream>
#include <alerror/alerror.h>
#include <alproxies/altexttospeechproxy.h>
#include <alproxies/alrobotpostureproxy.h>

#include <alcommon/alproxy.h>
#include <alcommon/albroker.h>

int main(int argc, char* argv[])
{
  if(argc != 4)
  {
    std::cerr << "Wrong number of arguments!" << std::endl;
    std::cerr << "Usage: setstate PEPPER_IP PEPPER_PORT posture" << std::endl;
    exit(2);
  }


  std::string pip = argv[1];
  int pport = atoi(argv[2]);//9559;
  std::string posture = argv[3];
  
  // A broker needs a name, an IP and a port to listen:
  const std::string brokerName = "mybroker";
  // Create your own broker
  boost::shared_ptr<AL::ALBroker> broker =
    AL::ALBroker::createBroker(brokerName, "0.0.0.0", 54000, pip, pport);

  // Create a proxy to ALTextToSpeechProxy
  AL::ALProxy proxyTTS(broker, "ALTextToSpeech");

  AL::ALRobotPostureProxy proxyRP(broker);
  std::string current_posture = proxyRP.getPosture();
  std::cerr << "Robot posture: " << current_posture << std::endl;
    
  /** The desired phrase to be said. */
  std::string phraseToSay = "Hello! My current posture is " + current_posture;
  // Call say method
  proxyTTS.callVoid("say", phraseToSay);
  
  std::vector<std::string> familypostures = proxyRP.getPostureFamilyList();
  for (size_t i=0; i<familypostures.size(); i++){
    std::string fposture = familypostures[i];
    std::cerr << "Posture Family: " << fposture << std::endl;
  }

  std::vector<std::string> postures = proxyRP.getPostureList();
  for (size_t i=0; i<postures.size(); i++){
    std::string posture = postures[i];
    std::cerr << "Possible postures: " << posture << std::endl;
  }


  if (current_posture != posture){
    phraseToSay = "Changing my posture to " + posture;
    proxyTTS.callVoid("say", phraseToSay);

    proxyRP.goToPosture(posture,1.0);

    current_posture = proxyRP.getPosture();
    std::cerr << "Robot posture: " << current_posture << std::endl;
  } else {
    phraseToSay = "Nothing to change here.";
    proxyTTS.callVoid("say", phraseToSay);
  }
  
  exit(0);
}
